### PR TITLE
Catch hydration errors and progressively hydrate array boundaries

### LINKED ISSUE #2539: bug when hydrating with error boundary
Hi, I am hydrating server rendered html and have an error boundary to handle client rendering errors. I am observing that if a component throws during hydration, both the original HTML and the "error" HTML are visible. You can reproduce the issue by opening the HTML below in a browser. You can uncomment the imports for react instead of preact to see the expected behavior.

### Steps to reproduce

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf8" />
  </head>
  <body>
    <div id="root">
      <div>Hello, World!</div>
    </div>
    <script type="module">
      // use these imports to observe incorrect behavior with preact
      import {
        Component,
        createElement,
        hydrate,
      } from "https://cdn.pika.dev/preact@10.4.1";
      import { useState } from "https://cdn.pika.dev/preact@10.4.1/hooks";

      // use these imports instead to observe correct behavior with react
      // import {
      //   Component,
      //   createElement,
      //   useState,
      // } from "https://cdn.pika.dev/react";
      // import { hydrate } from "https://cdn.pika.dev/react-dom";

      class Root extends Component {
        constructor() {
          super();
          this.state = {};
        }
        componentDidCatch(error) {
          this.setState({ error });
        }
        render() {
          if (!this.state.error) {
            return createElement(App);
          } else {
            return createElement("div", {}, "Error!");
          }
        }
      }

      function App() {
        throw Error(); // <-- throwing here does not work as expected
        return createElement("div", {}, "Hello, World!");
      }

      hydrate(createElement(Root), document.getElementById("root"));
    </script>
  </body>
</html>
```

### Expected Behavior

After hydration, "Hello, World!" should be replaced with "Error!" because the App component throws. When using the react imports instead of the preact imports in the file above, I see the expected behavior.

### Actual Behavior

I see both "Hello, World!" and "Error!" below it.

### PR DESCRIPTION
**Pull request by JoviDeCroock**
Fix error catching during hydration and progressive hydration of array-boundaries.

Fixes #2539

### CODE REVIEW

File: src/diff/index.js [hunk 1]
```diff
@@ -241,11 +241,19 @@ export function diff(
 		newVNode._original = null;
 		// if hydrating or creating initial tree, bailout preserves DOM:
 		if (isHydrating || excessDomChildren != null) {
-			newVNode._dom = oldDom;
-			newVNode._hydrating = !!isHydrating;
-			excessDomChildren[excessDomChildren.indexOf(oldDom)] = null;
-			// ^ could possibly be simplified to:
-			// excessDomChildren.length = 0;
+			if (e.then) {
```
Comment (JoviDeCroock) [inline src/diff/index.js:244]: This logic should only kick in when we are in a boundary, this means that only if we notice a promise is thrown up we should attempt at progressively hydrating this boundary.

File: src/diff/index.js [hunk 2]
```diff
@@ -241,11 +241,19 @@ export function diff(
...
+				newVNode._dom = oldDom;
+				newVNode._hydrating = !!isHydrating;
+				excessDomChildren.length = 0;
```
Comment (JoviDeCroock) [inline src/diff/index.js:247]: Formerly we did `excessDomChildren[excessDomChildren.indexOf(oldDom)] = null;` which didn't catch the case if the direct child of your throwing promise was a component that returned `x.map(Component)`.

File: src/diff/index.js [hunk 3]
```diff
@@ -241,11 +241,19 @@ export function diff(
...
+			} else {
+				// When we are hydrating and have excessDomChildren we don't know the _children this VNode
+				// should receive so it's safer to unmount them. Else on the subsequent error-boundary diff,
+				// we won't know the oldDom and insert an additional node instead of replace the prerendered one. (#2539)
+				for (tmp = excessDomChildren.length; tmp--; ) {
```
Comment (JoviDeCroock) [inline src/diff/index.js:252]: When we have an error being thrown during hydration we should clear the dom else we'll append the oldDom with the new one when we rerender.

File: src/diff/index.js (hunk 1 above)
Comment (developit) [reply src/diff/index.js:244]: Hmm - elsewhere we use component._dirty to trigger this - it's valid to throw a non-Promise and catch it, and that should trigger progressive hydration.

Comment (JoviDeCroock) [reply src/diff/index.js:244]: I mean, an error shouldn't hold progressive hydration. An error boundary will then trigger bugs like #2539

**Pull request closed without merging**